source/mesh_deformation/interface.cc

```diff
@@ -0,0 +1,639 @@
+
+      std::vector<types::global_dof_index> cell_dof_indices (dofs_per_cell);
+      std::vector<unsigned int> face_dof_indices (dofs_per_face);
+      Vector<double> cell_vector (dofs_per_cell);
+      FullMatrix<double> cell_matrix (dofs_per_cell, dofs_per_cell);
+
+      // We are just solving a Laplacian in each spatial direction, so
+      // the degrees of freedom for different dimensions do not couple.
+      Table<2,DoFTools::Coupling> coupling (dim, dim);
+      coupling.fill(DoFTools::none);
+
+      for (unsigned int c=0; c<dim; ++c)
+        coupling[c][c] = DoFTools::always;
+
+      LinearAlgebra::SparseMatrix mesh_matrix;
+#ifdef ASPECT_USE_PETSC
+      LinearAlgebra::DynamicSparsityPattern sp(mesh_locally_relevant);
+#else
+      TrilinosWrappers::SparsityPattern sp (mesh_locally_owned,
+                                            mesh_locally_owned,
+                                            mesh_locally_relevant,
+                                            sim.mpi_communicator);
+#endif
+      DoFTools::make_sparsity_pattern (free_surface_dof_handler,
+                                       coupling, sp,
+                                       mesh_displacement_constraints, false,
+                                       Utilities::MPI::
+                                       this_mpi_process(sim.mpi_communicator));
+#ifdef ASPECT_USE_PETSC
+      SparsityTools::distribute_sparsity_pattern(sp,
+                                                 free_surface_dof_handler.n_locally_owned_dofs_per_processor(),
+                                                 sim.mpi_communicator, mesh_locally_relevant);
+      sp.compress();
+      mesh_matrix.reinit (mesh_locally_owned, mesh_locally_owned, sp, sim.mpi_communicator);
+#else
+      sp.compress();
+      mesh_matrix.reinit (sp);
+#endif
+
+      // carry out the solution
+      FEValuesExtractors::Vector extract_vel(0);
+
+      LinearAlgebra::Vector rhs, velocity_solution;
+      rhs.reinit(mesh_locally_owned, sim.mpi_communicator);
+      velocity_solution.reinit(mesh_locally_owned, sim.mpi_communicator);
+
+      typename DoFHandler<dim>::active_cell_iterator cell = free_surface_dof_handler.begin_active(),
+                                                     endc= free_surface_dof_handler.end();
+      for (; cell!=endc; ++cell)
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (cell_dof_indices);
+            fe_values.reinit (cell);
+
+            cell_vector = 0;
+            cell_matrix = 0;
+            for (unsigned int point=0; point<n_q_points; ++point)
+              for (unsigned int i=0; i<dofs_per_cell; ++i)
+                {
+                  for (unsigned int j=0; j<dofs_per_cell; ++j)
+                    cell_matrix(i,j) += scalar_product( fe_values[extract_vel].gradient(j,point),
+                                                        fe_values[extract_vel].gradient(i,point) ) *
```

Comment:
reverse these two lines so that it is first `i` then `j`

Reply:
This only yields something nonzero if the components for `i` and `j` are the same. This could be checked above the computation here to skip the computation if possible.